compiler/noirc_evaluator/src/ssa/ir/instruction/call/blackbox.rs

```diff
@@ -0,0 +1,190 @@
+use std::rc::Rc;
+
+use acvm::{acir::AcirField, BlackBoxFunctionSolver, BlackBoxResolutionError, FieldElement};
+use iter_extended::vecmap;
+
+use crate::ssa::ir::{
+    dfg::DataFlowGraph, instruction::SimplifyResult, types::Type, value::ValueId,
+};
+
+use super::{array_is_constant, make_constant_array, to_u8_vec};
+
+pub(super) fn simplify_ec_add(
+    dfg: &mut DataFlowGraph,
+    solver: impl BlackBoxFunctionSolver<FieldElement>,
+    arguments: &[ValueId],
+) -> SimplifyResult {
+    match (
+        dfg.get_numeric_constant(arguments[0]),
+        dfg.get_numeric_constant(arguments[1]),
+        dfg.get_numeric_constant(arguments[2]),
+        dfg.get_numeric_constant(arguments[3]),
+        dfg.get_numeric_constant(arguments[4]),
+        dfg.get_numeric_constant(arguments[5]),
+    ) {
+        (
+            Some(point1_x),
+            Some(point1_y),
+            Some(point1_is_infinity),
+            Some(point2_x),
+            Some(point2_y),
+            Some(point2_is_infinity),
+        ) => {
+            let Ok((result_x, result_y, result_is_infinity)) = solver.ec_add(
+                &point1_x,
+                &point1_y,
+                &point1_is_infinity,
+                &point2_x,
+                &point2_y,
+                &point2_is_infinity,
+            ) else {
+                return SimplifyResult::None;
+            };
+
+            let result_x = dfg.make_constant(result_x, Type::field());
+            let result_y = dfg.make_constant(result_y, Type::field());
+            let result_is_infinity = dfg.make_constant(result_is_infinity, Type::bool());
+
+            let typ = Type::Array(Rc::new(vec![Type::field()]), 3);
+            let result_array =
+                dfg.make_array(im::vector![result_x, result_y, result_is_infinity], typ);
+
+            SimplifyResult::SimplifiedTo(result_array)
+        }
+        _ => SimplifyResult::None,
+    }
+}
+
+pub(super) fn simplify_poseidon2_permutation(
+    dfg: &mut DataFlowGraph,
+    solver: impl BlackBoxFunctionSolver<FieldElement>,
+    arguments: &[ValueId],
+) -> SimplifyResult {
+    match (dfg.get_array_constant(arguments[0]), dfg.get_numeric_constant(arguments[1])) {
+        (Some((state, _)), Some(state_length)) if array_is_constant(dfg, &state) => {
+            let state: Vec<FieldElement> = state
+                .iter()
+                .map(|id| {
+                    dfg.get_numeric_constant(*id)
+                        .expect("value id from array should point at constant")
+                })
+                .collect();
+
+            let Some(state_length) = state_length.try_to_u32() else {
+                return SimplifyResult::None;
+            };
+
+            let Ok(new_state) = solver.poseidon2_permutation(&state, state_length) else {
+                return SimplifyResult::None;
+            };
+
+            let result_array = make_constant_array(dfg, new_state, Type::field());
```

Comment:
nit: could re-use the type from `get_array_constant`, i.e. match on `Some((state, typ))` and then do
```suggestion
            let result_array = make_constant_array(dfg, new_state, typ);
```

Reply:
We'd end up with a doubly nested array type if we did this as `make_constant_array` wraps `typ` in `Type::Array`